docs/howtos/model_surgery.rst

```diff
@@ -0,0 +1,232 @@
+  from flax.core import unfreeze, freeze
+
+We will show how to get a flat dict of all the tensors, and then go back to a 
+nested, frozen dict. We show this for both Flax modules and optimizers.
+
+Surgery with Flax Modules
+--------------------------------
+
+Let's first make a little model for our demo.
+
+.. testcode::
+
+  class CNN(nn.Module):
+    @nn.compact
+    def __call__(self, x):
+      x = nn.Conv(features=32, kernel_size=(3, 3))(x)
+      x = nn.relu(x)
+      x = nn.avg_pool(x, window_shape=(2, 2), strides=(2, 2))
+      x = nn.Conv(features=64, kernel_size=(3, 3))(x)
+      x = nn.relu(x)
+      x = nn.avg_pool(x, window_shape=(2, 2), strides=(2, 2))
+      x = x.reshape((x.shape[0], -1))
+      x = nn.Dense(features=256)(x)
+      x = nn.relu(x)
+      x = nn.Dense(features=10)(x)
+      x = nn.log_softmax(x)
+      return x
+
+  def get_initial_params(rng):
+    init_shape = jnp.ones((1, 28, 28, 1), jnp.float32)
+    initial_params = CNN().init(rng, init_shape)['params']
+    return initial_params
+
+  key = random.PRNGKey(0)
+  params = get_initial_params(key)
+
+  print(jax.tree_map(jnp.shape, params))
+
+.. testoutput::
+
+  FrozenDict({
+      Conv_0: {
+          bias: (32,),
+          kernel: (3, 3, 1, 32),
+      },
+      Conv_1: {
+          bias: (64,),
+          kernel: (3, 3, 32, 64),
+      },
+      Dense_0: {
+          bias: (256,),
+          kernel: (3136, 256),
+      },
+      Dense_1: {
+          bias: (10,),
+          kernel: (256, 10),
+      },
+  })
+
+
+Let's get a flat dict for doing model surgery.
+
+.. testcode::
+
+  # Unfreeze params to normal dict.
+  params = unfreeze(params)
+  # Get flattened-key: value list.
+  flat_params = {'/'.join(k): v for k, v in traverse_util.flatten_dict(params).items()}
+  print(jax.tree_map(jnp.shape, flat_params))
+
+.. testoutput::
+  :options: +NORMALIZE_WHITESPACE
+
+  {'Conv_0/bias': (32,),
+   'Conv_0/kernel': (3, 3, 1, 32),
+   'Conv_1/bias': (64,),
+   'Conv_1/kernel': (3, 3, 32, 64),
+   'Dense_0/bias': (256,),
+   'Dense_0/kernel': (3136, 256),
+   'Dense_1/bias': (10,),
+   'Dense_1/kernel': (256, 10)}
+
+After doing whatever you want, unflatten back:
+
+.. testcode::
+
+  # Unflatten.
+  unflat_params = traverse_util.unflatten_dict({tuple(k.split('/')): v for k, v in flat_params.items()})
+  # Refreeze.
+  unflat_params = freeze(unflat_params)
+  print(jax.tree_map(jnp.shape, unflat_params))
+
+.. testoutput::
+  :options: +NORMALIZE_WHITESPACE
+
+  FrozenDict({
+      Conv_0: {
+          bias: (32,),
+          kernel: (3, 3, 1, 32),
+      },
+      Conv_1: {
+          bias: (64,),
+          kernel: (3, 3, 32, 64),
+      },
+      Dense_0: {
+          bias: (256,),
+          kernel: (3136, 256),
+      },
+      Dense_1: {
+          bias: (10,),
+          kernel: (256, 10),
+      },
+  })
+
+Surgey with Optimizers
+--------------------------------
+
+If you're loading from a flax optimizer, all of the variables live in
```

Comment:
it's only the variables that should be optimized by the optimizer, usually the `params` collection.